### ISSUE TITLE
Support for classic index replication setup with a self-contained Solr docker image

### ISSUE THREAD

**Issue by schuch** [querqy/querqy#872]
Our Solr setup looks like this:

- Classic index replication [^1]
- Self-contained Solr Docker image with prepared config sets as part of the image (readonly-fs). Only the Solr data directory is located on a writable volume.

It is therefore our understanding that we cannot use the `StandAloneRewriterContainer`[^3]. (cannot write the files to the `/conf` folder in the readonly config set). The `InMemoryRewriteContainer`[^4] does work for a single node, but what about replication? Also the rules would not be persistently stored.

**Possible Solution: "IndexRewriterContainer"**

We came up with the idea to implement our own `RewriterContainer` that stores the rewriter data in a dedicated Solr collection, just like Querqy Elasticsearch does.

Challenges:

1. If the configuration is stored in a Solr index, one must ensure that this collection is already loaded before the collection using the Querqy plugin tries to use it. We would need to implement a component, that is able to delay core startup if core it depends on is not yet ready.
1. When storing in the index, there is the challenge that the Querqy plugin must be communicated the rule updates on the followers so that the rules are reloaded. We think, that we can solve this with a `SolrEventListener`[^2]
1. The `RewriterContainer` is not extensible[^5][^6]. We would need to fork Querqy to achieve this solution, what we don't like.

**Questions**

- Can you think of any other solutions?
- Would it be possible to make the RewriterContainer pluggable to support our solution?

[^1]: [Classic Index Replication in the Solr reference guide](https://solr.apache.org/guide/solr/latest/deployment-guide/user-managed-index-replication.html)
[^2]: [SolrEventListener](https://solr.apache.org/docs/9_7_0/core/org/apache/solr/core/SolrEventListener.html) and [SolrCore#registerFirstSearcherListener](https://solr.apache.org/docs/9_7_0/core/org/apache/solr/core/SolrCore.html#registerFirstSearcherListener(org.apache.solr.core.SolrEventListener))
[^3]: [StandAloneRewriterContainer](https://github.com/querqy/querqy/blob/a6e99b968374c8a6977c7816ebd6fc636cb35555/querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/StandAloneRewriterContainer.java)
[^4]: [InMemoryRewriteContainer](https://github.com/querqy/querqy/blob/a6e99b968374c8a6977c7816ebd6fc636cb35555/querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/InMemoryRewriteContainer.java)
[^5]: `rewriteContainer` is `private` at [QuerqyRewriterRequestHandler.java#L152](https://github.com/querqy/querqy/blob/a6e99b968374c8a6977c7816ebd6fc636cb35555/querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/QuerqyRewriterRequestHandler.java#L152)
[^6]: Related discussion: https://github.com/querqy/querqy/pull/174#discussion_r603503679

**Comment by renekrie** [comment]
Hi @schuch, Thank you for bringing up this idea!

I'm trying to understand how the `SolrEventListener` would help with the replication. Is the idea that the listener would be configured for the Querqy config  index and get triggered on the follower if there were replicated data and then it would have to know which core(s) to inform for which rewriter updates and require that 2 data nodes must not use the same rewriter names?  (Under ES rewriters are being used independent of an index but under Solr the rewriter config endpoints are per index - we'd need to assure a consistent mapping between the rewriter config and the corresponding data index that uses the rewriter if we store the configs in an index)

Re   `rewriteContainer` is `private` at `QuerqyRewriterRequestHandler`: That's on purpose as the idea is to let `QuerqyRewriterRequestHandler` be the creator of the `rewriteContainer` instance in `inform(SolrCore)` based on the initArgs of the RequestHandler. I would keep it as such and not mix the decision based on initArgs with basing it on subclassing the RequestHandler. As far as I remember, we also used to have more `rewriteContainer instanceOf...` checks which in principle would be another argument for having an exhaustive knowledge about subclasses of `RewriteContainer` within the `QuerqyRewriterRequestHandler`. => I'd suggest you provide a PR that creates the instance in the `QuerqyRewriterRequestHandler`.

Blocking requests while the config index is not available yet: you could consider to still answer requests but without the rewriter being available. The `QuerqyQParserPlugin` has a config property `skipUnknownRewriters` which, if set to `true` would basically trigger that behaviour. (We'd basically say: that rewriter does not exist. The reason could be that it never was configured or that the configuration is not loaded yet.)

**Comment by mkr** [comment]
Yes, the `SolrEventListener` would run on the Querqy configuration core. It would need the core name(s) to inform on rule updates. It will then get the Querqy Request Handler from the core. We would need to get the `RewriterContainer` from the request handler and somehow refresh it / force it to reload all rewriter configurations (alternatively, we could do all the "work" of getting all the configuration documents and parsing them in this listener and just call `#saveRewriter` for each rewriter on a configured `InMemoryRewriterContainer`). We could add one document in the Querqy configuration core per core and rewriter (with id="<core>:<rewriterId>" and store the config as a single stored binary field.

**Comment by renekrie** [comment]
+1 for using binary as rewriter configs can easily grow into the MBs so that we might want to compress them.

I'd try to avoid putting knowledge about rewriters into the listener. That's what the container is for. So the listener should just notify the Querqy Request Handler (maybe just add an action "removeFromCache" that could also be implemented by the other  container implementations, or maybe action="refresh" where the InMemoryContainer could not support it)

**Comment by renekrie** [comment]
Hm, I wonder when exactly the "new searcher" even is triggered but you're probably right that the new searcher isn't available at that point apart from within the event listener.

The other not-so-nice thing might be that we'd have to reload all rewriter configurations, not just the ones that were updated (we wouldn't know from the listener what was updated). Maybe one alternative could be not to use a listener but to periodically check from within the RewriterContainer whether the config index has been updated and then figure out which configurations have been changed? A simple timestamp with ms precision might do - the container might keep track of the most recent timestamp that it has loaded and then only select docs that were updated after that? (maybe check every 15-30s?)

**Comment by schuch** [comment]
> Hm, I wonder when exactly the "new searcher" even is triggered but you're probably right that the new searcher isn't available at that point apart from within the event listener.

I had a little look into the Solr code and what I saw confirms the assumption that only the event listener sees the most recent searcher at the time of the event: The replication handler [forces a new searcher](https://github.com/apache/solr/blob/0c76b83de09c975c0ea2a8550b08c40ee9d15ad0/solr/core/src/java/org/apache/solr/handler/IndexFetcher.java#L1039C48-L1039C64). Within this action, the "new searcher" event [is propagated to the registered listeners](https://github.com/apache/solr/blob/0c76b83de09c975c0ea2a8550b08c40ee9d15ad0/solr/core/src/java/org/apache/solr/core/SolrCore.java#L2652).

I currently have a working solution with direct coupling between the event listener and the Query request handler. I plan to publish the code here shortly as a pull request (as soon as I have received internal approval for the contribution). Then we will have a basis for further discussion.

> The other not-so-nice thing might be that we'd have to reload all rewriter configurations, not just the ones that were updated

Yes, unfortunately the event listener cannot provide any information about the content of the change.
Do you think it is a big problem to do a full configuration reload?

The outlined pull strategy on the replicated configuration core seems to me to be a valid alternative.

**Comment by mkr** [comment]
> I currently have a working solution with direct coupling between the event listener and the Query request handler.

If we don't want to pass around the new searcher from the event listener to the RewriterContainer we could do _some_ more work in the listener, i.e. just use the searcher to retrieve all configuration documents and pass them (instead of the searcher) to a reload method in the RewriterContainer.

> Do you think it is a big problem to do a full configuration reload?

I think that also depends on the frequency / occasions this event is triggered. I would assume as this is on a dedicated core just for Querqy configuration documents new IndexSearchers are opened relatively rarely, i.e. only when something indeed has changed.

**Comment by renekrie** [comment]
Implemented
